### ISSUE TITLE
TypedDict undetected key error and type inference

### ISSUE THREAD

**Issue by marcofrasvda** [astral-sh/ty#2999]
### Summary

I've tried to use a TypedDict to prevent some mistakes but it seems ty allows them with no complaints. I see there's #154 but I feel these are pretty basic use cases and hopefully simple tests for when you get to implementing these checks.

```py
from typing import TypedDict

class JSONResponseItem(TypedDict):
    deviceId: int
    deviceName: str

class JSONResponse(TypedDict):
    success: bool
    items: list[JSONResponseItem]

def get_json():
    return {
        "success": True,
        "items": []
    }

def get_devices() -> None:
    json: JSONResponse = get_json()
    if json["sucscess"] + 3 == 5: #  <-- should signal a key error
        return
    for item in json["items"]:
        device_id = item["deviceId"] #  <-- Unknown, should be`int`
        device_name = item["deviceName"]  # <-- Unknown, should be`str`
    print("ok")
```

These behave as expected with Pyright and Pyrefly:
- I expect `ty` to signal that `sucscess` is not a valid key
- I expect `ty` to correctly infer the types when assigning the variables in the loop

This is a nice-to-have, currently not supported by other type checkers I've tried:
- I expect `ty` to signal that `+ 3` is not a valid operation on a `bool`

### Version

0.0.21

**Comment by AlexWaygood** [comment]
Thanks for the report!

The issue here is that your `get_json()` method is unannotated, so ty infers the `json` variable as `Unknown`, despite the fact that it's been explicitly annotated with `JsonResponse`. https://github.com/astral-sh/ty/issues/136 will let you use annotations to "upcast" dynamic types like `Any` or `Unknown` into fully static types, in the way you're trying to do here.

The other thing that might fix this for you is https://github.com/astral-sh/ty/issues/128, but I suspect the first version of that will not help you out here, because we'd need to synthesize a `TypedDict` for you when inferring the return type of your `get_json` function, which would be somewhat advanced.

**Comment by AlexWaygood** [comment]
> I expect `ty` to signal that `+ 3` is not a valid operation on a `bool`

Could you say a bit more about why you would like this? Unfortunately `bool` is a subclass of `int`, so it's very much valid from a type-theory perspective to treat it like an `int` and add `3` to an instance of `bool`:

```pycon
>>> isinstance(True, int)
True
>>> issubclass(bool, int)
True
>>> True + 3
4
>>> True == 1
True
>>> False == 0
True
```

**Comment by marcofrasvda** [comment]
Ooops, sorry I didn' t realize that specifying the variable type as `JSONResponse` was not enough to trigger the checks I needed, I though they were not implemented yet for some reason, thanks for linking to the relevant issues!

About the bool/int question, honestly I forgot that booleans are integers in Python, I see why it's technically valid but in practice I don't see a reason why one would want to add booleans and integers - it looks like an error and several statically typed languages (C#/Java/Go/Rust) don't allow this. In the code I shared, that I made up just to show my point, a warning for this would be helpful to indicate that perhaps the wrong dictionary item is being used for addition. I understand if this is not desirable, too niche or not in scope for ty, coming from C# I just thought this may be a source of bugs that can be catched by a type checker.

We could rephrase that point to be:
> I expect `ty` to warn that `+ 3` is not a recommended operation on a bool

P.S.: booleans were added in [2002](https://peps.python.org/pep-0285/) (24 years ago!) as integers for backwards compatibility, arguably math operations on booleans were not an intended use case but rather an unfortunate tradeoff

**Comment by AlexWaygood** [comment]
If I could go back in time and redesign Python from scratch, I definitely would not make `bool` a subclass of `int`! Unfortunately a lot of Python code depends on it being a subclass of `int` in 2026, however; code like this is common and idiomatic in Python:

```pycon
>>> results = ["pass", "fail", "fail", "pass", "fail", "fail"]
>>> num_passing = sum(result == "pass" for result in results)
>>> num_passing
2
```

**Comment by marcofrasvda** [comment]
Oh, first time I see that, thanks for sharing. I think this example specifically is compatible with the "no math operations on booleans" (_operators_ actually) idea, but I get your point. Looking forward to try the implementations of the issues you linked when you get to those!

**Comment by AlexWaygood** [comment]
Under the hood the reason why `sum` works with `bool` and the reason why `+` works with `bool` are the same: `bool` inherits the `__add__` method from its superclass `int`!
